core/lib/zksync_core/src/api_server/web3/state.rs

```diff
@@ -578,13 +582,29 @@ impl Filters {
 
     /// Retrieves filter from the state.
     pub fn get(&self, index: U256) -> Option<&TypedFilter> {
-        self.state.get(&index)
+        self.state.get(&index).map(|installed_filter| {
+            let previous_request_timestamp = installed_filter.last_request;
+            let now = chrono::Utc::now().naive_utc().timestamp() as u64;
+
+            *installed_filter.last_request = now;
+
+            let filter_type = FilterType::from(installed_filter.filter.clone());
+
+            FILTER_METRICS.request_frequency[&filter_type]
+                .observe(Duration::from_secs(now - previous_request_timestamp));
+
+            &installed_filter.filter
+        })
     }
 
     /// Updates filter in the state.
     pub fn update(&mut self, index: U256, new_filter: TypedFilter) -> bool {
         if let Some(typed_filter) = self.state.get_mut(&index) {
-            *typed_filter = new_filter;
+            let filter_type = FilterType::from(new_filter.clone());
+            let guard = FILTER_METRICS.metrics_count[&filter_type].inc_guard(1);
+
+            // FIXME: is previous value dropped?
+            *typed_filter = InstalledFilter::new(new_filter, guard);
```

Comment:
You shouldn't drop the previous value it'll lead to incorrect metrics, because semantically the filter is *updated*; it's still in place after the update. Instead, you should just update the `filter` field of the filter.